### ISSUE TITLE
Multiple connections created [API-236]

### ISSUE THREAD

**Issue by jerrinot** [hazelcast/hazelcast-csharp-client#375]
I have Hazelcast IMDG server running via Docker `$ docker run -p 5701:5701 hazelcast/hazelcast:latest-snapshot`
And this program: 
```c#
namespace helloHazelcast4
{
    class Program
    {
        public static async Task Main(string[] ignored)
        {
            var options = HazelcastOptions.Build();
            options.ClusterName = "dev";
            var hz = await HazelcastClientFactory.StartNewClientAsync(options, TimeSpan.MaxValue);
            await hz.DisposeAsync();
        }
    }
}
```
When start the program I see this in server logs:
```
2021-01-21 11:55:25,270 [ INFO] [hz.trusting_taussig.generic-operation.thread-0] [c.h.c.i.p.t.AuthenticationMessageTask]: [172.17.0.2]:5701 [dev] [4.2-SNAPSHOT] Received auth from Connection[id=1, /172.17.0.2:5701->/172.17.0.1:52106, qualifier=null, endpoint=[172.17.0.1]:52106, alive=true, connectionType=CSP, planeIndex=-1], successfully authenticated, clientUuid: bc34a169-9d0e-492f-afaa-e07caf32cfcb, client version: 4.0
2021-01-21 11:55:25,412 [ INFO] [hz.trusting_taussig.priority-generic-operation.thread-0] [c.h.c.i.p.t.AuthenticationMessageTask]: [172.17.0.2]:5701 [dev] [4.2-SNAPSHOT] Received auth from Connection[id=2, /172.17.0.2:5701->/172.17.0.1:43783, qualifier=null, endpoint=[172.17.0.1]:43783, alive=true, connectionType=CSP, planeIndex=-1], successfully authenticated, clientUuid: bc34a169-9d0e-492f-afaa-e07caf32cfcb, client version: 4.0
2021-01-21 11:55:25,471 [ INFO] [hz.trusting_taussig.IO.thread-in-1] [c.h.i.s.t.TcpServerConnection]: [172.17.0.2]:5701 [dev] [4.2-SNAPSHOT] Connection[id=2, /172.17.0.2:5701->/172.17.0.1:43783, qualifier=null, endpoint=[172.17.0.1]:43783, alive=false, connectionType=CSP, planeIndex=-1] closed. Reason: Connection closed by the other side
2021-01-21 11:55:25,489 [ INFO] [hz.trusting_taussig.IO.thread-in-0] [c.h.i.s.t.TcpServerConnection]: [172.17.0.2]:5701 [dev] [4.2-SNAPSHOT] Connection[id=1, /172.17.0.2:5701->/172.17.0.1:52106, qualifier=null, endpoint=[172.17.0.1]:52106, alive=false, connectionType=CSP, planeIndex=-1] closed. Reason: Connection closed by the other side
2021-01-21 11:55:31,831 [ INFO] [hz.trusting_taussig.cached.thread-1] [c.h.c.i.ClientEndpointManager]: [172.17.0.2]:5701 [dev] [4.2-SNAPSHOT] Destroying ClientEndpoint{connection=Connection[id=1, /172.17.0.2:5701->/172.17.0.1:52106, qualifier=null, endpoint=[172.17.0.1]:52106, alive=false, connectionType=CSP, planeIndex=-1], clientUuid='bc34a169-9d0e-492f-afaa-e07caf32cfcb, authenticated=true, clientVersion=4.0, creationTime=1611230125267, latest clientAttributes=null, labels=[]}
2021-01-21 11:55:31,831 [ INFO] [hz.trusting_taussig.cached.thread-1] [c.h.c.i.ClientEndpointManager]: [172.17.0.2]:5701 [dev] [4.2-SNAPSHOT] Destroying ClientEndpoint{connection=Connection[id=2, /172.17.0.2:5701->/172.17.0.1:43783, qualifier=null, endpoint=[172.17.0.1]:43783, alive=false, connectionType=CSP, planeIndex=-1], clientUuid='bc34a169-9d0e-492f-afaa-e07caf32cfcb, authenticated=true, clientVersion=4.0, creationTime=1611230125411, latest clientAttributes=null, labels=[]}
```
It appears as if the client created 2 connections. 

![image](https://user-images.githubusercontent.com/158619/105348110-210d6b80-5be8-11eb-8e3a-c17317a2bf8f.png)

**Comment by jerrinot** [comment]
I observe the same behaviour even with `hazelcast/hazelcast:4.1.1` instead of the `latest-snapshot`. that makes the reproducer more ... ~reproducible~ stable.

**Comment by zpqrtbnk** [comment]
Cannot repro on Windows, but can repro on Linux with the same very simple example, both with .NET Core 2.1 and 3.1. It *could* be that some things happen in a different order on Linux and we try to reconnect when we disconnect. Also note that this is reproduced with the `4.0.0-preview.5` version and we need to test with the very latest code.

Adding to the 4.0 milestone.

**Comment by zpqrtbnk** [comment]
Cause identified. Depending on the Docker setup, we connect to a member on 127.0.0.1:5701 but the member reports it runs on a different IP eg 172.17.0.2:5701 and that confuses us. Easy fix, fixing now.

**Comment by jerrinot** [comment]
@zdehz well done! 
are you sure the fix is easy? I think @sancar has a few horror stories regarding different IP/hostname resolution and matching.

**Comment by zpqrtbnk** [comment]
"Easy" might have been a bit quick. Here is more details:
* On Windows, we connect to 127.0.0.1:5701 and the member reports it runs on 172.17.0.2:5701 and we try to *also* connect to that address (yes, this is a mistake) and it fails = we have one connection
* On Linux, we connect to 127.0.0.1:5701 and the member reports it runs on 172.17.0.2:5701 and we try to *also* connect to that address (yes, this is a mistake) and it succeeds = we have two connections

**The mistake** (connecting to the member *twice* on different addresses) is easy to fix, that was a stupid error I did. **Fixing.**

**But I am wondering.**

Imagine we run several members with that configuration. On Windows, this apparently would not work since we don't seem to be able to reach the members at the address they report.

Is this
* An issue with how I configured Docker on Windows (we should be able to reach them at the address they report), not an Hazelcast problem
* An issue with how I configured *Hazelcast* itself (and then, what?)
